Data/Mod/Word.hs

```diff
@@ -345,22 +348,16 @@ half x = x `shiftR` 1
 -- >>> 4 ^% (-1) :: Mod 10 -- 4 and 10 are not coprime
 -- (*** Exception: divide by zero
 (^%) :: (KnownNat m, Integral a) => Mod m -> a -> Mod m
-mx@(Mod (W# x#)) ^% a = case natVal mx of
+mx@(Mod !x) ^% a = case natVal mx of
   NatJ#{} -> tooLargeModulus
-  NatS# m#
+  m@(NatS# _)
     | a < 0 -> case invertMod mx of
-      Nothing            -> throw DivideByZero
-      Just (Mod (W# y#)) -> Mod $ W# (f y# (- a) 1##)
-    | otherwise          -> Mod $ W# (f x# a 1##)
+      Nothing      -> throw DivideByZero
+      Just (Mod y) -> Mod $ f y (-a) 1
+    | otherwise    -> Mod $ f x a 1
     where
-      f :: Integral a => Word# -> a -> Word# -> Word#
-      f _  0 acc# = acc#
-      f b# e acc# = f bb# (e `P.quot` 2) (if odd e then ba# else acc#)
-        where
-          !(# bb1#, bb2# #) = timesWord2# b# b#
-          !(#    _, bb#  #) = quotRemWord2# bb1# bb2# m#
-          !(# ba1#, ba2# #) = timesWord2# b# acc#
-          !(#    _, ba#  #) = quotRemWord2# ba1# ba2# m#
+      f !_ 0 acc = acc
+      f b  e acc = f (mulMod m b b) (e `P.quot` 2) (if odd e then mulMod m b acc else acc)
```

Comment:
Are you sure that there is no performance regression?

Reply:
Pretty sure, the `Word`s still get unboxed and the generated core is even smaller (there are no dictionaries being created every iteration anymore). Additionally, `mulMod m b acc` is now only computed when it is needed.